ops/pebble.py

```diff
@@ -43,22 +43,243 @@
 import urllib.parse
 import urllib.request
 import warnings
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    BinaryIO,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Mapping,
+    Optional,
+    Sequence,
+    TextIO,
+    Tuple,
+    TypeVar,
+    Union,
+)
 
 from ops._private import yaml
 from ops._vendor import websocket
 
+if TYPE_CHECKING:
+    _Boundary = Tuple[int, int, bool]
+    from email.message import Message
+
+    from typing_extensions import Literal, Protocol, TypedDict
+
+    # callback types for _MultiParser header and body handlers
+    class _BodyHandler(Protocol):
+        def __call__(self, data: bytes, done: bool = False) -> None: ...  # noqa
+
+    _HeaderHandler = Callable[[bytes], None]
+    _StrOrByt = Union[str, bytes]
+    Numerical = Union[float, int]
+
+    # tempfile.NamedTemporaryFile has an odd interface because of that
+    # 'name' attribute, so we need to make a Protocol for it.
+    class _Tempfile(Protocol):
+        name = ''
+        def write(self, data: bytes): ...  # noqa
+        def close(self): ...  # noqa
+
+    class _FileLikeIO(Protocol[typing.AnyStr]):  # That also covers TextIO and BytesIO
+        def read(self, __n: int = ...) -> typing.AnyStr: ...  # for BinaryIO  # noqa
+        def write(self, __s: typing.AnyStr) -> int: ...  # noqa
+        def __enter__(self) -> typing.IO[typing.AnyStr]: ...  # noqa
+
+    class _Readable(Protocol):
+        def read(self, n: int = -1) -> _StrOrByt: ...  # noqa
+
+    class _Writable(Protocol):
+        # We'd need something like io.ReadableBuffer here,
+        # but we can't import that type
+        def write(self, buf: Union[bytes, str, bytearray]) -> int: ...  # noqa
```

Comment:
The types I added (excluding the inline ones and the TypedDicts) boil down to IO-related things.
Which was, to be honest, very difficult to type (and to understand how the objects are used, which attributes they're supposed to have, and so on. As @rwcarlsen pointed out (it was you, wasn't it?) that part of Python feels a bit messy.
I'm fine with `#ignore`ing all of that if you think it doesn't help, I thought it would help maintaining the code.
If in the future we feel that maintaining a specific signature is too much of a burden, we can always remove the type and `#ignore` away.

On the other hand, some say that complex types are a sign that a refactoring is due :)